python_modules/dagster/dagster/_core/storage/upath_io_manager.py

```diff
@@ -0,0 +1,195 @@
+        key_range = context.asset_partition_key_range
+        return key_range.start != key_range.end
+
+    def _get_path(self, context: Union[InputContext, OutputContext]) -> UPath:
+        if context.has_asset_key:
+            # we are dealing with an asset
+            context_path = ["assets"] + list(context.asset_key.path)
+
+            if context.has_asset_partitions:
+                # add partition
+                context_path += [context.partition_key]
+        else:
+            # we are dealing with an op output
+            context_path = ["runs"] + list(context.get_identifier())
+
+        return self._base_path.joinpath(*context_path).with_suffix(self.extension)
+
+    def load_input(self, context: InputContext) -> Union[Any, Dict[str, Any]]:
+        # In this load_input function, we vary the behavior based on the type of the downstream input
+        # if the type is a List, we load a list of mapped partitions.
+        path = self._get_path(context)
+        assert context.metadata is not None
+        allow_missing_partitions = context.metadata.get("allow_missing_partitions", False)
+
+        # unfortunately, this doesn't work for Python 3.7 : inspect.get_annotations(self.serialize)["obj"]
+        expected_type = inspect.signature(self.dump_to_path).parameters["obj"].annotation
+
+        if context.dagster_type.typing_type == expected_type and self._has_multiple_partitions(
+            context
+        ):
+            return check.failed(
+                f"Received `{context.dagster_type.typing_type.__name__}` type "
+                f"in input DagsterType {context.dagster_type}, "
+                f"but the input has multiple partitions. "
+                f"`Dict[str, {context.dagster_type.typing_type.__name__}]` should be used in this case."
+            )
+
+        if context.dagster_type.typing_type == expected_type and not self._has_multiple_partitions(
+            context
+        ):
+            context.log.debug(f"Loading from {path} using {self.__class__.__name__}")
+            obj = self.load_from_path(context=context, path=path)
+            context.add_input_metadata(
+                {
+                    "path": MetadataValue.path(path),
+                }
+            )
+            return obj
+
+        # FIXME: use a custom PartitionsDict type instead of Dict
+        elif (
+            hasattr(context.dagster_type.typing_type, "__origin__")
+            and context.dagster_type.typing_type.__origin__ in (Dict, dict)
+            and context.dagster_type.typing_type.__args__[1] == expected_type
+            and self._has_multiple_partitions(context)
+        ):
+            # load multiple partitions
+            if not context.has_asset_partitions:
+                raise TypeError(
+                    f"Detected {context.dagster_type.typing_type} input type "
+                    f"but the asset is not partitioned"
+                )
+
+            base_dir: UPath = path.parent
```

Comment:
The name `base_dir` is a tiny bit confusing here, because it resembles the `base_path` class attribute. Maybe `asset_dir` instead?